### PR TITLE
use model chunking on all paths

### PR DESCRIPTION
**Pull request by JensWehner**
This PR extends memory-aware model chunking from screened first-order XC evaluation to every relevant model execution path:

This reduces the memory consumption on all levels with 

- Applies model chunking to dense and screened XC energy/potential evaluation.
- Applies chunking to Hessian-response model evaluations.
- Applies chunking to CPU and GPU explicit nuclear-gradient VJPs.
- Retains full-grid evaluation for functionals that do not support spatial decomposition.
- It also simplifies the feature and integrator architecture:
- Introduces ModelFeaturePlan to separate packed AO requirements from model-visible features.
- Uses FeatureSpec.supports_spatial_decomposition as the single chunking capability signal.
- Consolidates dense and screened integration and response paths.

### CODE REVIEW

File: skala/src/skala/pyscf/features.py [hunk 1]
```diff
@@ -10,83 +10,8 @@
...
-            gpu=gpu,
-        )
-
-        for feature in mgga_features:
-            mol_features[feature] = feature_math.maybe_expand_and_divide(
-                mgga_features[feature], not is_spin_polarized, 2
-            )
-
-    return mol_features
+from skala.pyscf.evaluation import FeatureSpec
```
[the remark this answers is not in the document]
Comment (JensWehner) [reply skala/src/skala/pyscf/features.py:14]: I am okay with breaking the API

File: skala/src/skala/pyscf/gradients.py [hunk 2]
```diff
@@ -83,79 +44,28 @@ def veff_and_expl_nuc_grad(
...
-        ao = torch.from_numpy(
-            dft.numint.eval_ao(
-                mol, coords, deriv=ao_deriv, non0tab=mask, cutoff=grid.cutoff
+    ao_deriv, derivatives = evaluate_nuclear_feature_derivatives(
+        functional,
+        mol,
+        grid_,
+        rdm1,
+        nuc_grad_feats,
+        max_memory_in_mb=2000,
```
[the remark this answers is not in the document]
Comment (JensWehner) [reply skala/src/skala/pyscf/gradients.py]: fixed

File: pixi.lock
Comment (JensWehner) [inline pixi.lock:1]: These are only minor version upgrades + the addition of pyspy.

File: skala/src/skala/pyscf/gradient_core.py [hunk 3]
```diff
@@ -146,6 +125,116 @@ def grid_derivative_block(
...
+        Feature.ATOMIC_GRID_SIZE_BOUND_SHAPE,
+    }
+    unsupported = differentiable_features - SUPPORTED_NUCLEAR_GRADIENT_FEATURES
+    if unsupported:
+        raise NotImplementedError(
+            f"Not supported features for nuclear gradient: {unsupported}"
+        )
+
+    ao_features = differentiable_features & AO_FEATURES
+    ao_deriv = ao_derivative_order(ao_features) + int(bool(ao_features))
```
Comment (JensWehner) [inline skala/src/skala/pyscf/gradient_core.py]: explain the second term in a comment

File: skala/src/skala/pyscf/model_chunking.py [hunk 4]
```diff
@@ -121,12 +210,40 @@ class ModelFeatureChunk:
...
-class ModelFeatureChunker:
-    """Prepared atom-aligned partition of raw and model features.
+@dataclass(frozen=True)
+class ModelFeaturePlan:
+    """Describe feature evaluation and model inputs for model chunking.
 
-    The chunker is a snapshot of the supplied density matrix, grid, and raw
-    features. It can be iterated repeatedly while those inputs represent the
-    same calculation, but must not be reused after their state changes.
+    The evaluation specification contains every feature needed by the calculation,
```
Comment (JensWehner) [inline skala/src/skala/pyscf/model_chunking.py:217]: add examples and maybe a post init that makese sure that model_feature_spec is a subset of evaluation_feature_sepc

File: skala/src/skala/pyscf/model_chunking.py [hunk 5]
```diff
@@ -121,12 +210,40 @@ class ModelFeatureChunk:
...
+    """
+
+    evaluation_feature_spec: FeatureSpec
+    model_feature_spec: FeatureSpec
+
+
+class ModelFeatureChunker:
+    """Prepare atom-aligned model inputs from globally evaluated AO features.
+
+    ``atom_major_raw_features`` must use the packed channel layout derived from
```
Comment (JensWehner) [inline skala/src/skala/pyscf/model_chunking.py]: delete paragraph

File: skala/src/skala/pyscf/model_chunking.py [hunk 6]
```diff
@@ -135,94 +252,230 @@ def __init__(
...
+    """Differentiate a scalar XC energy with respect to molecular features."""
+    if not feature_tensors:
+        return ()
+
+    differentiable_features = tuple(
+        tensor.detach().requires_grad_(True) for tensor in feature_tensors
+    )
+    exc = exc_func(*differentiable_features)
+    if not exc.requires_grad:
+        return tuple(torch.zeros_like(feature) for feature in differentiable_features)
```
Comment (JensWehner) [inline skala/src/skala/pyscf/model_chunking.py]: we wanted to raise here instead.

Comment (JensWehner) [reply skala/src/skala/pyscf/model_chunking.py]: or I am not sure if only for gradient.

File: skala/src/skala/pyscf/model_chunking.py [hunk 7]
```diff
@@ -135,94 +252,230 @@ def __init__(
...
+    )
+    gradients = {
+        feature_name: torch.zeros_like(model_features[feature_name])
+        for feature_name in differentiable_features
+    }
+    if not differentiable_features:
+        return gradients
+
+    for indices in chunk_indices:
+        chunk_features: FeatureMap = {}
```
Comment (JensWehner) [inline skala/src/skala/pyscf/model_chunking.py:418]: Sometimes I think FeatureMap could be a class that handles some of this functionality itself

File: skala/tests/test_xc_integrator.py [hunk 8]
```diff
@@ -87,22 +89,43 @@ def test_xc_integrator_returns_tensors_and_xc_only_response(
...
     grids = SkalaGrids(mol)
 
-    def fake_generate_features(
-        mol: gto.Mole,
+    def fake_evaluate_raw_features(
         dm: torch.Tensor,
+        mol: gto.Mole,
         grids: object,
-        features: set[Feature],
+        feature_function: object,
```
Comment (JensWehner) [inline skala/tests/test_xc_integrator.py]: type properly

File: skala/src/skala/pyscf/model_chunking.py [hunk 9]
```diff
@@ -135,94 +269,200 @@ def __init__(
...
+    ao_features = feature_spec.ao_features
+    nfeatures = ao_features.nfeats if ao_features is not None else 0
+    chunk_indices = _make_model_chunk_indices(
+        dm,
+        atomic_grid_sizes,
+        nfeatures=nfeatures,
+        deriv_order=1,
+        max_memory_in_mb=max_memory_in_mb,
+        safety_fraction=safety_fraction,
+    )
```
Comment (jhrmnn) [inline skala/src/skala/pyscf/model_chunking.py]: This can be moved below under the `not differential_features` early return

File: pixi.lock
Comment (jhrmnn) [reply pixi.lock]: Doesn't pixi support "minimal update" of an env to satisfy updated constraints? (to avoid the "minor version upgrades"

File: skala/src/skala/pyscf/feature_math.py [hunk 10]
```diff
@@ -78,50 +200,36 @@ def forward(self, dm: torch.Tensor, ao: torch.Tensor) -> torch.Tensor:
...
+                ci = dm_view @ ao_gradient
+                gradient_contraction += torch.sum(ci * ao_gradient[None], dim=-2)
+
+        for feature, feature_slice in self._feature_spec:
+            if feature == Feature.DENSITY:
+                feature_values = torch.sum(c0 * phi[None], dim=-2).unsqueeze(-2)
+            elif feature == Feature.GRAD:
+                feature_values = 2 * torch.sum(
+                    c0[:, None] * packed_ao.gradient[None], dim=-2
                 )
```
Comment (jhrmnn) [inline skala/src/skala/pyscf/feature_math.py]: This now takes 3 times as much memory as the previous code which had it unrolled over the cartesian axes, is that intentional?

**Review by awvwgk — APPROVED** [review]
Looks good to me.
